### ISSUE TITLE
Ghost workspace icons with steam

### ISSUE THREAD

**Issue by noah427** [JakeStanger/ironbar#533]
**Describe the bug**
opening and closing the steam window causes the workspace module to leave behind ghost workspaces.

**To Reproduce**
1. Open steam
2. Close steam window
3. Change workspaces
4. Repeat to make infinite ghost workspace icons

**Expected behavior**
the workspace icon disappears when the steam window is no longer in it

**System information:**
 - Distro: Nixos
- Compositor: Hyprland
 - Ironbar version: ironbar 0.15.0-pre

**Configuration**

[config.json](https://github.com/JakeStanger/ironbar/files/14892981/config.json)

**Videos**

https://github.com/JakeStanger/ironbar/assets/40678304/bfb19919-6503-430f-829c-3c191372e639

https://github.com/JakeStanger/ironbar/assets/40678304/3e851ae9-bbe2-4ed1-bf72-a25f0a470ffd

**Comment by yavko** [comment]
Unless for some reason steam leaves open clients on that workspace, which prevent the workspace close event, I don't know what else it could be

**Comment by JakeStanger** [comment]
There's something funny going on and it's not Steam. You can see in your first video you manage to open a duplicate `3` workspace almost immediately, but I can't easily reproduce that. It's like the existing workspace button is out of sync and shouldn't be there?

If you restart Ironbar, are you able to easily reproduce that in particular? If you are, can you also provide logs with `IRONBAR_LOG=debug` please?

**Comment by noah427** [comment]
I can reproduce it every single time I restart ironbar. Here's a log file I recorded while I recreated the issue. Hopefully this sheds some light on the problem
[log.txt](https://github.com/JakeStanger/ironbar/files/14895424/log.txt)

**Comment by JakeStanger** [comment]
Right, bear with me because I'm struggling to replicate it and I'm still trying to make sense of the issue.

I've mapped out the workspace state changes below. Workspaces being added/removed happens before/after the focus event respectively, which is why it's a bit verbose. 

The 2nd time that you move to workspace 4, it looks like workspace 3 never receives a destroy event. 

```
[1, 2*]
[1*, 2]
[1, 2*]
[1, 2*, 3]
[1, 2, 3*]
[1, 2*, 3]
[1, 2*]
[1, 2*, 3]
[1, 2, 3*]
[1, 2, 3*, 4]
[1, 2, 3, 4*]
[1, 2, 3*, 4]
[1, 2, 3*]
[1, 2, 3*, 4]
[1, 2, 3, 4*] <-- here, why is workspace 3 sticking around?
[1, 2*, 3, 4]
[1, 2*, 3]
```

That would indicate Ironbar is *probably* behaving correctly, as it's interpreting the events it receives and is never told to remove the client.

I would suggest the next step is to find out what the Hyprland event socket is actually sending. You can do something like this:

```sh
socat - "UNIX-CONNECT:/tmp/hypr/$HYPRLAND_INSTANCE_SIGNATURE/.socket2.sock" | while read -r line; do echo "$line"; done
````

The following events are of particular interest:

- `focusedmon` - includes workspace focus change if you change focused output
- `createworkspace`
- `workspace` - focusing a workspace on the same output
- `destroyworkspace` - this is the one you want to watch especially

Also, I could see in your log that each workspace event is received twice. Normally, that'd indicate two running workspace modules (ie two bars on two different screens), but the log suggested you only have a single monitor. Any idea what's going on there?

**Comment by noah427** [comment]
I actually do have two bars, one on each monitor. Not sure why the logs don't show that. I'll check on the hyprland event socket later today and report back with what I find. My current theory is that there's some issue with applications that spawn floating tiles because that's the pattern I've noticed with the problems so far. Thanks for all your help so far!

**Comment by JakeStanger** [comment]
> Not sure why the logs don't show that.

Ah they do to some extent, but every workspace event is logged on DP-1 and it suggests no workspaces ever exist on DP-2. Which seems wrong and odd. That could be a bug with the actual log code...

**Comment by noah427** [comment]
Here's what I see while reproducing this issue with the socket open
```
openlayer>>wofi
closelayer>>wofi
activewindow>>,
activewindowv2>>,
activewindow>>foot,noah@noah: ~
activewindowv2>>2cfb390
activewindow>>steam,Steam
activewindowv2>>2cbef40
openwindow>>2cbef40,14,steam,Steam
closewindow>>2cbef40
activewindow>>foot,noah@noah: ~
activewindowv2>>2cfb390
```

**Comment by JakeStanger** [comment]
So in this case you were just switching between workspace 1 and 2? I think I've interpreted the workspace changes as follows:

```
[1*]
[1*, 2]
[1, 2*]
[1*, 2]
[1*]
[1*, 2]
[1, 2*]
[1*, 2]
[1*]
[1*, 2]
[1, 2*]
```

(beginning to feel like some advent of code input...)

The log here suggests all is fine. What I might need to do then is try some small-scale test and see if we can pin down exactly where in the pipeline the issue is.

It's damn annoying I can't replicate this myself too.

**Comment by JakeStanger** [comment]
Going to close this issue as I've had confirmation from @noah427 that various system updates have resolved this
